connector/avro/src/main/scala/org/apache/spark/sql/avro/AvroDeserializer.scala

```diff
@@ -117,6 +119,19 @@ private[sql] class AvroDeserializer(
     val incompatibleMsg = errorPrefix +
         s"schema is incompatible (avroType = $avroType, sqlType = ${catalystType.sql})"
 
+    val realDataType = SchemaConverters.toSqlType(avroType).dataType
+    val confKey = SQLConf.LEGACY_AVRO_ALLOW_READING_WITH_INCOMPATIBLE_SCHEMA
+    val preventReadingIncorrectType = !SQLConf.get.getConf(confKey)
+    def isNotExactType: Boolean = {
+      preventReadingIncorrectType && !DataTypeUtils.sameType(realDataType, catalystType)
```

Comment:
what if we do the check and fail immediately here? It seems can make the code simpler.

Reply:
I don't think I can do that. We still need to allow reading decimal format with higher precision

Reply:
We can special-case decimal type in the check
```
(realDataType, catalystType) match {
  case (DecimalType(p1, s1), DecimalType(p2, s2)) => check precision and scale
  case => preventReadingIncorrectType && !DataTypeUtils.sameType(realDataType, catalystType)
}
```

Reply:
Yeah, I use 
```
(avroType.getType, catalystType) match {
  // check whether allow conversion of DecimalType
  case (FIXED, dt: DecimalType) =>
    val d = avroType.getLogicalType.asInstanceOf[LogicalTypes.Decimal]
    if (preventReadingIncorrectType &&
      d.getPrecision - d.getScale > dt.precision - dt.scale) {
      throw lowerPrecisionException(dt)
    }
  case (BYTES, dt: DecimalType) =>
    val d = avroType.getLogicalType.asInstanceOf[LogicalTypes.Decimal]
    if (preventReadingIncorrectType &&
      d.getPrecision - d.getScale > dt.precision - dt.scale) {
      throw lowerPrecisionException(dt)
    }
  // show allow following conversion
  case (RECORD, _: StructType) =>
  case (LONG, _: TimestampType)
    if !preventReadingIncorrectType || !realDataType.isInstanceOf[DayTimeIntervalType] =>
  case (LONG, _: TimestampNTZType)
    if !preventReadingIncorrectType || !realDataType.isInstanceOf[DayTimeIntervalType] =>
  // check whether allow different type
  case _ if isNotExactType =>
    throw incorrectTypeException(catalystType)
  case _ =>
}
```
to fail earier. I am not sure whether it makes the code simpler. What do you think?

Reply:
We need to exclude a lot of other conversions. I don't think it's more elegant than the original one.

Reply:
I see. Since there are so many special cases, it's better to use the original code style and only check the data type for certain cases.

Reply:
sounds good. Let me revert this change!